webportal/src/app/cluster-view/hardware/hardware.component.js

```diff
@@ -17,14 +17,153 @@
 
 
 // module dependencies
+require('datatables.net/js/jquery.dataTables.js');
+require('datatables.net-bs/js/dataTables.bootstrap.js');
+require('datatables.net-bs/css/dataTables.bootstrap.css');
+require('datatables.net-plugins/sorting/natural.js');
+require('datatables.net-plugins/sorting/ip-address.js');
+require('jquery.ajax-cross-origin/js/jquery.ajax-cross-origin.min.js')
+const url = require('url');
 const hardwareComponent = require('./hardware.component.ejs');
+const breadcrumbComponent = require('../../job/breadcrumb/breadcrumb.component.ejs');
+const loading = require('../../job/loading/loading.component');
 const webportalConfig = require('../../config/webportal.config.json');
 
 const hardwareHtml = hardwareComponent({
+  breadcrumb: breadcrumbComponent,
   grafanaUri: webportalConfig.grafanaUri
 });
 
-$("#sidebar-menu--cluster-view").addClass("active");
-$("#sidebar-menu--cluster-view--hardware").addClass("active");
+function createCORSRequest(method, url) {
+  var xhr = new XMLHttpRequest();
+  if ("withCredentials" in xhr) {
+    // Check if the XMLHttpRequest object has a "withCredentials" property.
+    // "withCredentials" only exists on XMLHTTPRequest2 objects.
+    xhr.open(method, url, true);
+  } else if (typeof XDomainRequest != "undefined") {
+    // Otherwise, check if XDomainRequest.
+    // XDomainRequest only exists in IE, and is IE's way of making CORS requests.
+    xhr = new XDomainRequest();
+    xhr.open(method, url);
+  } else {
+    // Otherwise, CORS is not supported by the browser.
+    xhr = null;
+  }
+  return xhr;
+}
 
-$('#content-wrapper').html(hardwareHtml);
\ No newline at end of file
+
+const loadMachines = () => {
```

Comment:
@abuccts : Hi Yifan. I tried various ways to call kubernetes api in javascript here but all of them failed. The browser blocks the cross-domain call even if I sent a CORS request, since the server seemed have not enabled 'Access-Control-Allow-Origin'. Do you know the correct way to do it? I guess the 'cors' package you involved in rest-server is just for server side, right? Thanks!

Reply:
you may need to config cors in kubernetes side